File: speedy-test-lc/python/pipeline.py
"""Simple data-processing pipeline with composable stages."""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Callable, Generic, Iterable, Iterator, TypeVar

T = TypeVar("T")
U = TypeVar("U")


@dataclass
class Stage(Generic[T, U]):
    name: str
    fn: Callable[[T], U]

    def __call__(self, value: T) -> U:
        return self.fn(value)


@dataclass
class Pipeline(Generic[T]):
    stages: list[Stage] = field(default_factory=list)

    def pipe(self, name: str, fn: Callable) -> "Pipeline":
        self.stages.append(Stage(name=name, fn=fn))
        return self

    def run(self, source: Iterable[T]) -> Iterator:
        for item in source:
            result = item
            for stage in self.stages:
                result = stage(result)
            yield result

    def run_list(self, source: Iterable[T]) -> list:
        return list(self.run(source))


def words_pipeline() -> Pipeline:
    return (
        Pipeline()
        .pipe("strip",    str.strip)
        .pipe("lower",    str.lower)
        .pipe("tokenise", str.split)
        .pipe("filter",   lambda tokens: [t for t in tokens if len(t) > 2])
    )


if __name__ == "__main__":
    lines = [
        "  The quick brown fox  ",
        "Jumps Over the Lazy Dog",
        "a to be or not to be",
    ]
    pl = words_pipeline()
    for tokens in pl.run(lines):
        print(tokens)

File: speedy-test-lc/scala/Pipeline.scala
package speedytest

import scala.collection.mutable

/** Immutable transformation pipeline over a lazy stream of records. */
final class Pipeline[A] private (private val stages: List[A => Option[A]]) {

  def filter(p: A => Boolean): Pipeline[A] =
    new Pipeline(stages :+ (a => if (p(a)) Some(a) else None))

  def map[B >: A](f: A => B): Pipeline[B] =
    new Pipeline[B](stages.asInstanceOf[List[B => Option[B]]] :+ (b => Some(f(b.asInstanceOf[A]).asInstanceOf[B])))

  def run(source: Iterable[A]): List[A] =
    source.foldLeft(mutable.ListBuffer.empty[A]) { (acc, item) =>
      stages.foldLeft(Option(item)) { (opt, stage) => opt.flatMap(stage) }
        .foreach(acc += _)
      acc
    }.toList
}

object Pipeline {
  def of[A]: Pipeline[A] = new Pipeline(Nil)
}

// Example: word-frequency counter
object WordCount extends App {
  val lines = List(
    "the quick brown fox",
    "the fox jumped over the lazy dog",
    "dogs and foxes",
  )

  val words = Pipeline
    .of[String]
    .filter(_.nonEmpty)
    .run(lines.flatMap(_.split("\\s+")))

  val freq = words.groupMapReduce(identity)(_ => 1)(_ + _)
  freq.toSeq.sortBy(-_._2).take(5).foreach { case (w, n) => println(s"$w: $n") }
}
